crates/nu-protocol/src/syntax_shape.rs

```diff
@@ -97,9 +97,29 @@ pub enum SyntaxShape {
 
     /// A custom shape with custom completion logic
     Custom(Box<SyntaxShape>, DeclId),
+
+    /// Nothing
+    Nothing,
 }
 
 impl SyntaxShape {
+    pub fn accepts_cell_paths(&self) -> bool {
+        matches!(
+            self,
+            SyntaxShape::List(_) | SyntaxShape::Record | SyntaxShape::Table
+        )
+    }
+
+    pub fn is_subtype(&self, other: &SyntaxShape) -> bool {
+        match (self, other) {
+            (t, u) if t == u => true,
+            (SyntaxShape::Int, SyntaxShape::Number) => true,
+            (_, SyntaxShape::Any) => true,
+            (SyntaxShape::List(t), SyntaxShape::List(u)) if t.is_subtype(u) => true, // List is covariant
+            _ => false,
+        }
+    }
```

Comment:
cc @jntrnr I've started modeling the subtype relation here (but using `SyntaxShape` since that's what this PR uses for now).